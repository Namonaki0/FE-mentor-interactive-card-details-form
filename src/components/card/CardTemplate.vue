<script setup lang="ts">
import { computed } from 'vue'
import { useCardDetailsStore } from '@/stores/cardDetails'

const cardStore = useCardDetailsStore()

const props = defineProps<{
  cardfacet?: string
}>()

const cardImage = computed(() => {
  return props.cardfacet === 'back'
    ? new URL(`@/assets/images/bg-card-back.png`, import.meta.url).href
    : new URL(`@/assets/images/bg-card-front.png`, import.meta.url).href
})

const cardLogo = new URL('@/assets/images/card-logo.svg', import.meta.url).href
</script>

<template>
  <div class="card-wrapper">
    <img :src="cardImage" alt="Card Image" class="card-image" />
    <div class="cvc-wrapper" v-if="props.cardfacet === 'back'">
      <p>{{ cardStore.cvv || '000' }}</p>
    </div>
    <div class="card-details" v-else>
      <div class="logo-wrapper">
        <img :src="cardLogo" alt="Card Logo" class="card-logo" />
      </div>
      <div class="card-data-wrapper">
        <p class="card-number-data">{{ cardStore.formattedCardNumber || '0000 0000 0000 0000' }}</p>
        <div class="card-name-date">
          <p class="card-name">{{ cardStore.cardName || 'Jane Appleseed' }}</p>
          <p class="card-date">
            {{ (cardStore.expiryDateMonth || '00') + '/' + (cardStore.expiryDateYear || '00') }}
          </p>
        </div>
      </div>
    </div>
  </div>
</template>

<style lang="scss" scoped>
.card-wrapper {
  position: relative;
  width: clamp(300px, 75%, 400px);

  .card-image {
    width: 100%;
    display: block;
  }

  .card-details {
    .logo-wrapper {
      position: absolute;
      top: 12%;
      left: 7%;
      width: 25%;
      height: auto;

      .card-logo {
        width: 100%;
        display: block;
      }
    }

    .card-data-wrapper {
      position: absolute;
      bottom: 10%;
      left: 7%;
      font-size: 12px;
      color: white;
      width: calc(100% - 14%);

      .card-number-data {
        font-size: 1.23rem;
        letter-spacing: 1px;

        @media (min-width: $breakpoint-tablet) {
          font-size: 1.5rem;
          letter-spacing: 3px;
        }
      }

      .card-name-date {
        display: flex;
        justify-content: space-between;
      }
    }
  }

  .cvc-wrapper {
    position: absolute;
    top: 42%;
    right: 12%;
    height: 15%;
    display: flex;
    align-items: center;
    justify-content: flex-end;
    color: white;

    p {
      font-size: 14px;
      margin-bottom: 3px;
    }
  }
}
</style>
