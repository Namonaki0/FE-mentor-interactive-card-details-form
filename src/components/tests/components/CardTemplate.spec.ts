import { render, screen } from '@testing-library/vue'
import CardTemplate from '@/components/card/CardTemplate.vue'
import { createTestingPinia } from '@pinia/testing'
import { describe, it, expect } from 'vitest'

const renderCardTemplate = (facet = 'front', cardState = {}) => {
  return render(CardTemplate, {
    props: { cardfacet: facet },
    global: {
      plugins: [
        createTestingPinia({
          initialState: {
            cardDetails: {
              cardName: 'Andre F.',
              cardNumber: '1234 5678 1234 5678',
              expiryDateMonth: '08',
              expiryDateYear: '29',
              cvc: '999',
              ...cardState,
            },
          },
          stubActions: false,
        }),
      ],
    },
  })
}

describe('CardTemplate.vue', () => {
  it('renders front facet with correct card info', async () => {
    renderCardTemplate('front')

    expect(screen.getByText(/andre f\./i)).toBeInTheDocument()
    expect(screen.getByText(/1234 5678 1234 5678/)).toBeInTheDocument()
    expect(screen.getByText(/08\/29/)).toBeInTheDocument()
  })

  it('renders back facet with cvc', async () => {
    renderCardTemplate('back')

    expect(screen.getByText(/999/)).toBeInTheDocument()
    expect(screen.queryByText(/andre f\./i)).not.toBeInTheDocument()
    expect(screen.queryByText(/08\/29/)).not.toBeInTheDocument()
  })

  it('matches snapshot', () => {
    const { container } = renderCardTemplate('front')
    expect(container).toMatchSnapshot()
  })

  it('renders fallback values when store is empty', () => {
    renderCardTemplate('front', {
      cardName: '',
      cardNumber: '',
      expiryDateMonth: '',
      expiryDateYear: '',
      cvc: '',
    })

    expect(screen.getByText('Jane Appleseed')).toBeInTheDocument()
    expect(screen.getByText('0000 0000 0000 0000')).toBeInTheDocument()
    expect(screen.getByText('00/00')).toBeInTheDocument()
  })
})
