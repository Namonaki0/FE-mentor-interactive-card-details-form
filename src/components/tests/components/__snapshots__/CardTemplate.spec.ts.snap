// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CardTemplate.vue > matches snapshot 1`] = `
<div>
  <div
    class="card-wrapper"
    data-v-6bd020e7=""
  >
    <img
      alt="Card Image"
      class="card-image"
      data-v-6bd020e7=""
      src="http://localhost:3000/src/assets/images/bg-card-front.png"
    />
    <div
      class="card-details"
      data-v-6bd020e7=""
    >
      <div
        class="logo-wrapper"
        data-v-6bd020e7=""
      >
        <img
          alt="Card Logo"
          class="card-logo"
          data-v-6bd020e7=""
          src="data:image/svg+xml,%3csvg%20width='84'%20height='47'%20fill='none'%20xmlns='http://www.w3.org/2000/svg'%3e%3cellipse%20cx='23.478'%20cy='23.5'%20rx='23.478'%20ry='23.5'%20fill='%23fff'/%3e%3cpath%20d='M83.5%2023.5c0%205.565-4.507%2010.075-10.065%2010.075-5.559%200-10.065-4.51-10.065-10.075%200-5.565%204.506-10.075%2010.065-10.075%205.558%200%2010.065%204.51%2010.065%2010.075Z'%20stroke='%23fff'/%3e%3c/svg%3e"
        />
      </div>
      <div
        class="card-data-wrapper"
        data-v-6bd020e7=""
      >
        <p
          class="card-number-data"
          data-v-6bd020e7=""
        >
          1234 5678 1234 5678
        </p>
        <div
          class="card-name-date"
          data-v-6bd020e7=""
        >
          <p
            class="card-name"
            data-v-6bd020e7=""
          >
            Andre F.
          </p>
          <p
            class="card-date"
            data-v-6bd020e7=""
          >
            08/29
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
